modules/dnn/src/onnx/onnx_graph_simplifier.cpp

```diff
@@ -145,13 +154,258 @@ class SoftMaxSubgraph : public Subgraph
     int axis;
 };
 
+class FuseUpsampleSubgraph : public Subgraph
+{
+public:
+    FuseUpsampleSubgraph() : scaleH(1), scaleW(1), emptyNodeId(-1) {}
+
+    void finalize(const Ptr<ImportGraphWrapper>& net,
+                  const Ptr<ImportNodeWrapper>& fusedNode,
+                  std::vector<Ptr<ImportNodeWrapper> >& inputs) CV_OVERRIDE
+    {
+        opencv_onnx::NodeProto* node = fusedNode.dynamicCast<ONNXNodeWrapper>()->node;
+        opencv_onnx::AttributeProto* attrH = node->add_attribute();
+        attrH->set_name("height_scale");
+        attrH->set_i(scaleH);
+        opencv_onnx::AttributeProto* attrW = node->add_attribute();
+        attrW->set_name("width_scale");
+        attrW->set_i(scaleW);
+        //remove empty constant node from the net
+        if (emptyNodeId != -1)
+            net->removeNode(emptyNodeId);
+    }
+
+protected:
+    int scaleH, scaleW, emptyNodeId;
+};
+
+class ExtractScalesSubgraph : public FuseUpsampleSubgraph
+{
+public:
+    ExtractScalesSubgraph()
+    {
+        int input = addNodeToMatch("");
+
+        int indexH = addNodeToMatch("Constant");
+        int shape1 = addNodeToMatch("Shape", input);
+        int gather1 = addNodeToMatch("Gather", shape1, indexH);
+        int castG1 = addNodeToMatch("Cast", gather1);
+        int scaleH = addNodeToMatch("Constant");
+        int mul1 = addNodeToMatch("Mul", castG1, scaleH);
+        int castM1 = addNodeToMatch("Cast", mul1);
+        int floor1 = addNodeToMatch("Floor", castM1);
+
+        int indexW = addNodeToMatch("Constant");
+        int shape2 = addNodeToMatch("Shape", input);
+        int gather2 = addNodeToMatch("Gather", shape2, indexW);
+        int castG2 = addNodeToMatch("Cast", gather2);
+        int scaleW = addNodeToMatch("Constant");
+        int mul2 = addNodeToMatch("Mul", castG2, scaleW);
+        int castM2 = addNodeToMatch("Cast", mul2);
+        int floor2 = addNodeToMatch("Floor", castM2);
+
+        int unsqueeze1 = addNodeToMatch("Unsqueeze", floor1);
+        int unsqueeze2 = addNodeToMatch("Unsqueeze", floor2);
+        addNodeToMatch("Concat", unsqueeze1, unsqueeze2);
+
+        setFusedNode("Scales", input);
+    }
```

Comment:
Please take a look at the changes from https://github.com/dkurt/opencv/commit/15132e85b3f19782b1f40019877196a1331c34ca

I just refactored this part to be more object oriented. So to  fuse each graph we will have not a sequence of fusions but a single pattern.

There are also some improvements which let to manage empty Constant nodes or other stuff.

Reply:
@dkurt I have gone through all of your changes, this is much better way of doing fusion. My changes were really redundant. I have added your suggested changes. I will keep in mind to minimize and optimize code and make it object oriented as much as possible from now onwards. Thanks for your help.

Reply:
@ashishkrshrivastava, Anyway it's really interesting idea to combine multiple patters, really. I believe it can be useful somewhere but in this sample it's better to keep them separately for now.

Reply:
Sure.